// @microsoft/mgt-components/dist/es6/components/mgt-tasks-base/mgt-tasks-base.js
/**
 * -------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.
 * See License in the project root for license information.
 * -------------------------------------------------------------------------------------------
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { html, property } from 'lit-element';
import { classMap } from 'lit-html/directives/class-map';
import { Providers, ProviderState, MgtTemplatedComponent } from '@microsoft/mgt-element';
import { strings } from './strings';
/**
 * The foundation for creating task based components.
 *
 * @export
 * @class MgtTasksBase
 * @extends {MgtTemplatedComponent}
 */
export class MgtTasksBase extends MgtTemplatedComponent {
    constructor() {
        super();
        this.clearState();
        this._previousMediaQuery = this.mediaQuery;
        this.onResize = this.onResize.bind(this);
    }
    /**
     * The name of a potential new task
     *
     * @readonly
     * @protected
     * @type {string}
     * @memberof MgtTasksBase
     */
    get newTaskName() {
        return this._newTaskName;
    }
    get strings() {
        return strings;
    }
    /**
     * Synchronizes property values when attributes change.
     *
     * @param {*} name
     * @param {*} oldValue
     * @param {*} newValue
     * @memberof MgtTasks
     */
    attributeChangedCallback(name, oldVal, newVal) {
        super.attributeChangedCallback(name, oldVal, newVal);
        switch (name) {
            case 'target-id':
            case 'initial-id':
                this.clearState();
                this.requestStateUpdate();
                break;
        }
    }
    /**
     * updates provider state
     *
     * @memberof MgtTasks
     */
    connectedCallback() {
        super.connectedCallback();
        window.addEventListener('resize', this.onResize);
    }
    /**
     * removes updates on provider state
     *
     * @memberof MgtTasks
     */
    disconnectedCallback() {
        window.removeEventListener('resize', this.onResize);
        super.disconnectedCallback();
    }
    /**
     * Invoked on each update to perform rendering tasks. This method must return
     * a lit-html TemplateResult. Setting properties inside this method will *not*
     * trigger the element to update.
     */
    render() {
        const provider = Providers.globalProvider;
        if (!provider || provider.state !== ProviderState.SignedIn) {
            return html ``;
        }
        if (this.isLoadingState) {
            return this.renderLoadingTask();
        }
        const headerTemplate = !this.hideHeader ? this.renderHeader() : null;
        const newTaskTemplate = this._isNewTaskVisible ? this.renderNewTaskPanel() : null;
        const tasksTemplate = this.isLoadingState ? this.renderLoadingTask() : this.renderTasks();
        return html `
      ${headerTemplate} ${newTaskTemplate}
      <div class="Tasks" dir=${this.direction}>
        ${tasksTemplate}
      </div>
    `;
    }
    /**
     * Render the header part of the component.
     *
     * @protected
     * @returns
     * @memberof MgtTodo
     */
    renderHeader() {
        const headerContentTemplate = this.renderHeaderContent();
        const addClasses = classMap({
            AddBarItem: true,
            NewTaskButton: true,
            hidden: this.readOnly || this._isNewTaskVisible
        });
        return html `
      <div class="Header" dir=${this.direction}>
        ${headerContentTemplate}
        <button class="${addClasses}" @click="${() => this.showNewTaskPanel()}">
          <span class="TaskIcon"></span>
          <span>${this.strings.addTaskButtonSubtitle}</span>
        </button>
      </div>
    `;
    }
    /**
     * Render a task in a loading state.
     *
     * @protected
     * @returns
     * @memberof MgtTodo
     */
    renderLoadingTask() {
        return html `
      <div class="Task LoadingTask">
        <div class="TaskContent">
          <div class="TaskCheckContainer">
            <div class="TaskCheck"></div>
          </div>
          <div class="TaskDetailsContainer">
            <div class="TaskTitle"></div>
            <div class="TaskDetails">
              <span class="TaskDetail">
                <div class="TaskDetailIcon"></div>
                <div class="TaskDetailName"></div>
              </span>
              <span class="TaskDetail">
                <div class="TaskDetailIcon"></div>
                <div class="TaskDetailName"></div>
              </span>
            </div>
          </div>
        </div>
      </div>
    `;
    }
    /**
     * Render the panel for creating a new task
     *
     * @protected
     * @returns {TemplateResult}
     * @memberof MgtTasksBase
     */
    renderNewTaskPanel() {
        const newTaskName = this._newTaskName;
        const taskTitle = html `
      <input
        type="text"
        placeholder="${this.strings.newTaskPlaceholder}"
        .value="${newTaskName}"
        label="new-taskName-input"
        aria-label="new-taskName-input"
        role="input"
        @input="${(e) => {
            this._newTaskName = e.target.value;
            this.requestUpdate();
        }}"
      />
    `;
        const taskAddClasses = classMap({
            Disabled: !this._isNewTaskBeingAdded && (!newTaskName || !newTaskName.length),
            TaskAddButtonContainer: true
        });
        const taskAddTemplate = !this._isNewTaskBeingAdded
            ? html `
          <div class="TaskIcon TaskCancel" @click="${() => this.hideNewTaskPanel()}">
            <span>${this.strings.cancelNewTaskSubtitle}</span>
          </div>
          <div class="TaskIcon TaskAdd" @click="${() => this.addTask()}">
            <span></span>
          </div>
        `
            : null;
        const newTaskDetailsTemplate = this.renderNewTaskDetails();
        return html `
      <div dir=${this.direction} class="Task NewTask Incomplete">
        <div class="TaskContent">
          <div class="TaskDetailsContainer">
            <div class="TaskTitle">
              ${taskTitle}
            </div>
            <div class="TaskDetails">
              ${newTaskDetailsTemplate}
            </div>
          </div>
        </div>
        <div class="${taskAddClasses}">
          ${taskAddTemplate}
        </div>
      </div>
    `;
    }
    /**
     * Render a bucket icon.
     *
     * @protected
     * @returns
     * @memberof MgtTodo
     */
    renderBucketIcon() {
        return html `
      <svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
        <path
          fill-rule="evenodd"
          clip-rule="evenodd"
          d="M14 2H2V4H3H5H6H10H11H13H14V2ZM10 5H6V6H10V5ZM5 5H3V14H13V5H11V6C11 6.55228 10.5523 7 10 7H6C5.44772 7 5 6.55228 5 6V5ZM1 5H2V14V15H3H13H14V14V5H15V4V2V1H14H2H1V2V4V5Z"
          fill="#3C3C3C"
        />
      </svg>
    `;
    }
    /**
     * Create a new todo task and add it to the list
     *
     * @protected
     * @returns
     * @memberof MgtTasksBase
     */
    addTask() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this._isNewTaskBeingAdded || !this.newTaskName) {
                return;
            }
            this._isNewTaskBeingAdded = true;
            yield this.requestUpdate();
            try {
                yield this.createNewTask();
            }
            finally {
                this._isNewTaskBeingAdded = false;
                this._isNewTaskVisible = false;
                this.requestUpdate();
            }
        });
    }
    /**
     * Clear the form data from the new task panel.
     *
     * @protected
     * @memberof MgtTasksBase
     */
    clearNewTaskData() {
        this._newTaskName = '';
    }
    /**
     * Clear the component state.
     *
     * @protected
     * @memberof MgtTasksBase
     */
    clearState() {
        this.clearNewTaskData();
        this._isNewTaskVisible = false;
        this.requestUpdate();
    }
    /**
     * Handle when a task is clicked
     *
     * @protected
     * @param {Event} e
     * @param {TodoTask} task
     * @memberof MgtTasksBase
     */
    handleTaskClick(e, task) {
        this.fireCustomEvent('taskClick', { task });
    }
    /**
     * Convert a date to a properly formatted string
     *
     * @protected
     * @param {Date} date
     * @returns
     * @memberof MgtTasksBase
     */
    dateToInputValue(date) {
        if (date) {
            return new Date(date.getTime() - date.getTimezoneOffset() * 60000).toISOString().split('T')[0];
        }
        return null;
    }
    showNewTaskPanel() {
        this._isNewTaskVisible = true;
        this.requestUpdate();
    }
    hideNewTaskPanel() {
        this._isNewTaskVisible = false;
        this.clearNewTaskData();
        this.requestUpdate();
    }
    onResize() {
        if (this.mediaQuery !== this._previousMediaQuery) {
            this._previousMediaQuery = this.mediaQuery;
            this.requestUpdate();
        }
    }
}
__decorate([
    property({ attribute: 'read-only', type: Boolean }),
    __metadata("design:type", Boolean)
], MgtTasksBase.prototype, "readOnly", void 0);
__decorate([
    property({ attribute: 'hide-header', type: Boolean }),
    __metadata("design:type", Boolean)
], MgtTasksBase.prototype, "hideHeader", void 0);
__decorate([
    property({ attribute: 'hide-options', type: Boolean }),
    __metadata("design:type", Boolean)
], MgtTasksBase.prototype, "hideOptions", void 0);
__decorate([
    property({ attribute: 'target-id', type: String }),
    __metadata("design:type", String)
], MgtTasksBase.prototype, "targetId", void 0);
__decorate([
    property({ attribute: 'initial-id', type: String }),
    __metadata("design:type", String)
], MgtTasksBase.prototype, "initialId", void 0);
